app/src/main/java/com/github/se/icebreakrr/model/location/LocationViewModel.kt

```diff
@@ -0,0 +1,135 @@
+
+class LocationViewModel(
+    private val locationService: LocationService,
+    private val geoFirestoreRepository: GeoFirestoreRepository,
+    private val permissionManager: PermissionManager
+) : ViewModel() {
+
+  private val _isUpdatingLocation = MutableStateFlow(false)
+  val isUpdatingLocation: StateFlow<Boolean>
+    get() = _isUpdatingLocation
+
+  private var permissionObserverJob: Job? = null
+
+  companion object {
+    fun provideFactory(
+        locationService: LocationService,
+        geoFirestoreRepository: GeoFirestoreRepository,
+        permissionManager: PermissionManager
+    ): ViewModelProvider.Factory =
+        object : ViewModelProvider.Factory {
+          @Suppress("UNCHECKED_CAST")
+          override fun <T : ViewModel> create(modelClass: Class<T>): T {
+            if (modelClass.isAssignableFrom(LocationViewModel::class.java)) {
+              return LocationViewModel(locationService, geoFirestoreRepository, permissionManager)
+                  as T
+            }
+            throw IllegalArgumentException("Unknown ViewModel class")
+          }
+        }
+  }
+
+  /**
+   * Starts location updates if `ACCESS_FINE_LOCATION` permission is granted.
+   *
+   * Checks if location updates are already active. If not, verifies permission status:
+   * - Requests permission if not granted.
+   * - Starts location updates if permission is granted.
+   *
+   * Begins observing permission changes to manage updates accordingly.
+   */
+  fun tryToStartLocationUpdates() {
+    if (_isUpdatingLocation.value) return
+    val permission = Manifest.permission.ACCESS_FINE_LOCATION
+    val permissionStatus = permissionManager.permissionStatuses.value[permission]
+    if (permissionStatus != PackageManager.PERMISSION_GRANTED) {
+      permissionManager.requestPermissions(arrayOf(permission))
+    } else startServiceLocationUpdates()
+    if (permissionObserverJob == null) observePermissions()
+  }
+
+  /** Stops location updates and stops observing permission status changes. */
+  fun stopLocationUpdates() {
+    if (_isUpdatingLocation.value) {
+      locationService.stopLocationUpdates()
+      stopObservingPermissions()
+      _isUpdatingLocation.value = false
+    }
+  }
+
+  /**
+   * Initiates location updates from `LocationService` and updates Firestore with the user's
+   * position.
+   *
+   * Converts each new location into a `GeoPoint` and saves it in Firestore through
+   * `geoFirestoreRepository`. Sets `_isUpdatingLocation` to `true` if updates start successfully,
+   * or `false` if an error occurs.
+   */
+  private fun startServiceLocationUpdates() {
+    val locationUpdatesStarted =
+        locationService.startLocationUpdates(
+            onLocationUpdate = { location ->
+              val geoPoint = GeoPoint(location.latitude, location.longitude)
+              geoFirestoreRepository.setUserPosition(
+                  geoPoint) // Call Firestore update on location change
+            })
```

Comment:
Nitpick: I see `startLocationUpdates` has a parameter "onError", shouldn't you use it here to handle errors more gracefully?

Reply:
I've added a onError() parameter if we want to handle errors a different way in the future, for now just a error log is shown.